[BUG] Multi-thread flusher causes data inconsistent with chunk offsets.

### What is the bug?
A clear and concise description of what the bug is.

### How to reproduce the bug?
This problem happens randomly. There are no clear procedures to reproduce.

### Could you share logs or screenshots?

> Caused by: java.lang.IndexOutOfBoundsException: readerIndex(25350231) + length(30480) exceeds writerIndex(25358446): UnpooledDuplicatedByteBuf(ridx: 25350231, widx: 25358446, cap: 25358446, unwrapped: CompositeByteBuf(ridx: 13, widx: 25358446, cap: 25358446, components=454))
>   at com.aliyun.emr.io.netty.buffer.AbstractByteBuf.checkReadableBytes0(AbstractByteBuf.java:1442)

--- a/server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/FileWriter.java
+++ b/server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/FileWriter.java
@@ -61,7 +61,7 @@ public final class FileWriter extends DeviceObserver {
   private long bytesFlushed;
 
   private final DiskFlusher flusher;
-  private final int flusherReplicationIndex;
+  private final int flushWorkerIndex;
   private CompositeByteBuf flushBuffer;
 
   private final long chunkSize;
@@ -123,7 +123,7 @@ public final class FileWriter extends DeviceObserver {
     PartitionType partitionType) throws IOException {
     this.file = file;
     this.flusher = flusher;
-    this.flusherReplicationIndex = flusher.getReplicationIndex();
+    this.flushWorkerIndex = flusher.getWorkerIndex();
     this.dataRootDir = workingDir;
     this.chunkSize = chunkSize;
     this.nextBoundary = chunkSize;
@@ -321,7 +321,7 @@ public final class FileWriter extends DeviceObserver {
     }
 
     // real action
-    flushBuffer = flusher.takeBuffer(timeoutMs, flusherReplicationIndex);
+    flushBuffer = flusher.takeBuffer(timeoutMs, flushWorkerIndex);
 
     // metrics end
     if (source.samplePerfCritical()) {
@@ -336,7 +336,7 @@ public final class FileWriter extends DeviceObserver {
   }
 
   private void addTask(FlushTask task) throws IOException {
-    if (!flusher.addTask(task, timeoutMs, flusherReplicationIndex)) {
+    if (!flusher.addTask(task, timeoutMs, flushWorkerIndex)) {
       IOException e = new IOException("Add flush task timeout.");
       notifier.setException(e);
       throw e;
@@ -345,7 +345,7 @@ public final class FileWriter extends DeviceObserver {
 
   private synchronized void returnBuffer() {
     if (flushBuffer != null) {
-      flusher.returnBuffer(flushBuffer, flusherReplicationIndex);
+      flusher.returnBuffer(flushBuffer, flushWorkerIndex);
       flushBuffer = null;
     }
   }

--- a/server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/FileWriter.java
+++ b/server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/FileWriter.java
@@ -61,6 +61,7 @@ public final class FileWriter extends DeviceObserver {
   private long bytesFlushed;
 
   private final DiskFlusher flusher;
+  private final int flusherReplicationIndex;
   private CompositeByteBuf flushBuffer;
 
   private final long chunkSize;
@@ -122,6 +123,7 @@ public final class FileWriter extends DeviceObserver {
     PartitionType partitionType) throws IOException {
     this.file = file;
     this.flusher = flusher;
+    this.flusherReplicationIndex = flusher.getReplicationIndex();
     this.dataRootDir = workingDir;
     this.chunkSize = chunkSize;
     this.nextBoundary = chunkSize;
@@ -319,7 +321,7 @@ public final class FileWriter extends DeviceObserver {
     }
 
     // real action
-    flushBuffer = flusher.takeBuffer(timeoutMs);
+    flushBuffer = flusher.takeBuffer(timeoutMs, flusherReplicationIndex);
 
     // metrics end
     if (source.samplePerfCritical()) {
@@ -334,7 +336,7 @@ public final class FileWriter extends DeviceObserver {
   }
 
   private void addTask(FlushTask task) throws IOException {
-    if (!flusher.addTask(task, timeoutMs)) {
+    if (!flusher.addTask(task, timeoutMs, flusherReplicationIndex)) {
       IOException e = new IOException("Add flush task timeout.");
       notifier.setException(e);
       throw e;
@@ -343,7 +345,7 @@ public final class FileWriter extends DeviceObserver {
 
   private synchronized void returnBuffer() {
     if (flushBuffer != null) {
-      flusher.returnBuffer(flushBuffer);
+      flusher.returnBuffer(flushBuffer, flusherReplicationIndex);
       flushBuffer = null;
     }
   }

--- a/server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala
+++ b/server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala
@@ -56,11 +56,7 @@ private[worker] final class DiskFlusher(
   private val workingQueues = new Array[LinkedBlockingQueue[FlushTask]](threadCount)
   private val bufferQueues = new Array[LinkedBlockingQueue[CompositeByteBuf]](threadCount)
   for (index <- 0 until threadCount) {
-    val actualQueueSize = if (queueCapacity / threadCount > 0) {
-      queueCapacity / threadCount
-    } else {
-      100
-    }
+    val actualQueueSize = queueCapacity / threadCount + 1
     workingQueues(index) = new LinkedBlockingQueue[FlushTask](actualQueueSize)
     bufferQueues(index) = new LinkedBlockingQueue[CompositeByteBuf](actualQueueSize)
     for (_ <- 0 until actualQueueSize) {
@@ -68,7 +64,7 @@ private[worker] final class DiskFlusher(
     }
   }
   private val workers = new Array[Thread](threadCount)
-  private val nextReplicationIndex = new AtomicInteger()
+  private val nextWorkerIndex = new AtomicInteger()
 
   @volatile
   private var lastBeginFlushTime: Long = -1
@@ -119,24 +115,28 @@ private[worker] final class DiskFlusher(
     deviceMonitor.registerDiskFlusher(this)
   }
 
-  def getReplicationIndex: Int = {
-    nextReplicationIndex.getAndIncrement() % threadCount
+  def getWorkerIndex: Int = {
+    val nextIndex = nextWorkerIndex.getAndIncrement()
+    if (nextIndex > 1000000) {
+      nextWorkerIndex.set(0)
+    }
+    nextIndex % threadCount
   }
 
-  def takeBuffer(timeoutMs: Long, replicationIndex: Int): CompositeByteBuf = {
-    bufferQueues(replicationIndex).poll(timeoutMs, TimeUnit.MILLISECONDS)
+  def takeBuffer(timeoutMs: Long, workerIndex: Int): CompositeByteBuf = {
+    bufferQueues(workerIndex).poll(timeoutMs, TimeUnit.MILLISECONDS)
   }
 
-  def returnBuffer(buffer: CompositeByteBuf, replicationIndex: Int): Unit = {
+  def returnBuffer(buffer: CompositeByteBuf, workerIndex: Int): Unit = {
     MemoryTracker.instance().releaseDiskBuffer(buffer.readableBytes())
     buffer.removeComponents(0, buffer.numComponents())
     buffer.clear()
 
-    bufferQueues(replicationIndex).put(buffer)
+    bufferQueues(workerIndex).put(buffer)
   }
 
-  def addTask(task: FlushTask, timeoutMs: Long, replicationIndex: Int): Boolean = {
-    workingQueues(replicationIndex).offer(task, timeoutMs, TimeUnit.MILLISECONDS)
+  def addTask(task: FlushTask, timeoutMs: Long, workerIndex: Int): Boolean = {
+    workingQueues(workerIndex).offer(task, timeoutMs, TimeUnit.MILLISECONDS)
   }
 
   override def notifyError(deviceName: String, dirs: ListBuffer[File] = null,
@@ -150,10 +150,10 @@ private[worker] final class DiskFlusher(
         logError(s"Exception when interrupt worker: $workers, $e")
     }
     workingQueues.foreach { queue =>
-      queue.asScala.foreach(task => {
+      queue.asScala.foreach { task =>
         task.buffer.removeComponents(0, task.buffer.numComponents())
         task.buffer.clear()
-      })
+      }
     }
     deviceMonitor.unregisterDiskFlusher(this)
   }

--- a/server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala
+++ b/server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala
@@ -53,13 +53,22 @@ private[worker] final class DiskFlusher(
   val deviceMonitor: DeviceMonitor,
   val threadCount: Int) extends DeviceObserver with Logging {
   private lazy val diskFlusherId = System.identityHashCode(this)
-  private val workingQueue = new LinkedBlockingQueue[FlushTask](queueCapacity)
-  private val bufferQueue = new LinkedBlockingQueue[CompositeByteBuf](queueCapacity)
-  private val writeActionPool = ThreadUtils.newDaemonFixedThreadPool(threadCount,
-    workingDir.getName + "-flusher")
-  for (_ <- 0 until queueCapacity) {
-    bufferQueue.put(Unpooled.compositeBuffer(256))
+  private val workingQueues = new Array[LinkedBlockingQueue[FlushTask]](threadCount)
+  private val bufferQueues = new Array[LinkedBlockingQueue[CompositeByteBuf]](threadCount)
+  for (index <- 0 until threadCount) {
+    val actualQueueSize = if (queueCapacity / threadCount > 0) {
+      queueCapacity / threadCount
+    } else {
+      100
+    }
+    workingQueues(index) = new LinkedBlockingQueue[FlushTask](actualQueueSize)
+    bufferQueues(index) = new LinkedBlockingQueue[CompositeByteBuf](actualQueueSize)
+    for (_ <- 0 until actualQueueSize) {
+      bufferQueues(index).put(Unpooled.compositeBuffer(256))
+    }
   }
+  private val workers = new Array[Thread](threadCount)
+  private val nextReplicationIndex = new AtomicInteger()
 
   @volatile
   private var lastBeginFlushTime: Long = -1
@@ -68,12 +77,14 @@ private[worker] final class DiskFlusher(
   val stopFlag = new AtomicBoolean(false)
   val rand = new Random()
 
-  private val worker = new Thread(s"$this") {
-    override def run(): Unit = {
-      while (!stopFlag.get()) {
-        val task = workingQueue.take()
-        writeActionPool.submit(new Runnable {
-          override def run(): Unit = {
+  init()
+
+  private def init(): Unit = {
+    for (index <- 0 until (threadCount)) {
+      workers(index) = new Thread(s"$this-$index") {
+        override def run(): Unit = {
+          while (!stopFlag.get()) {
+            val task = workingQueues(index).take()
             val key = s"DiskFlusher-$workingDir-${rand.nextInt()}"
             workerSource.sample(WorkerSource.FlushDataTime, key) {
               if (!task.notifier.hasException) {
@@ -90,38 +101,42 @@ private[worker] final class DiskFlusher(
                 }
                 lastBeginFlushTime = -1
               }
-              returnBuffer(task.buffer)
+              returnBuffer(task.buffer, index)
               task.notifier.numPendingFlushes.decrementAndGet()
             }
           }
-        })
+        }
       }
+      workers(index).setDaemon(true)
+      workers(index).setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler {
+        override def uncaughtException(t: Thread, e: Throwable): Unit = {
+          logError(s"$this thread terminated.", e)
+        }
+      })
+      workers(index).start()
     }
-  }
-  worker.setDaemon(true)
-  worker.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler {
-    override def uncaughtException(t: Thread, e: Throwable): Unit = {
-      logError(s"$this thread terminated.", e)
-    }
-  })
-  worker.start()
 
-  deviceMonitor.registerDiskFlusher(this)
-
-  def takeBuffer(timeoutMs: Long): CompositeByteBuf = {
-    bufferQueue.poll(timeoutMs, TimeUnit.MILLISECONDS)
+    deviceMonitor.registerDiskFlusher(this)
   }
 
-  def returnBuffer(buffer: CompositeByteBuf): Unit = {
+  def getReplicationIndex: Int = {
+    nextReplicationIndex.getAndIncrement() % threadCount
+  }
+
+  def takeBuffer(timeoutMs: Long, replicationIndex: Int): CompositeByteBuf = {
+    bufferQueues(replicationIndex).poll(timeoutMs, TimeUnit.MILLISECONDS)
+  }
+
+  def returnBuffer(buffer: CompositeByteBuf, replicationIndex: Int): Unit = {
     MemoryTracker.instance().releaseDiskBuffer(buffer.readableBytes())
     buffer.removeComponents(0, buffer.numComponents())
     buffer.clear()
 
-    bufferQueue.put(buffer)
+    bufferQueues(replicationIndex).put(buffer)
   }
 
-  def addTask(task: FlushTask, timeoutMs: Long): Boolean = {
-    workingQueue.offer(task, timeoutMs, TimeUnit.MILLISECONDS)
+  def addTask(task: FlushTask, timeoutMs: Long, replicationIndex: Int): Boolean = {
+    workingQueues(replicationIndex).offer(task, timeoutMs, TimeUnit.MILLISECONDS)
   }
 
   override def notifyError(deviceName: String, dirs: ListBuffer[File] = null,
@@ -129,16 +144,17 @@ private[worker] final class DiskFlusher(
     logError(s"$this is notified Device $deviceName Error $deviceErrorType! Stop Flusher.")
     stopFlag.set(true)
     try {
-      worker.interrupt()
-      writeActionPool.shutdown()
+      workers.foreach(_.interrupt())
     } catch {
       case e: Exception =>
-        logError(s"Exception when interrupt worker: $worker, $e")
+        logError(s"Exception when interrupt worker: $workers, $e")
     }
-    workingQueue.asScala.foreach(task => {
-      task.buffer.removeComponents(0, task.buffer.numComponents())
-      task.buffer.clear()
-    })
+    workingQueues.foreach { queue =>
+      queue.asScala.foreach(task => {
+        task.buffer.removeComponents(0, task.buffer.numComponents())
+        task.buffer.clear()
+      })
+    }
     deviceMonitor.unregisterDiskFlusher(this)
   }
 
@@ -147,7 +163,7 @@ private[worker] final class DiskFlusher(
     deviceMonitor.reportDeviceError(workingDir, e, deviceErrorType)
   }
 
-  def bufferQueueInfo(): String = s"$this available buffers: ${bufferQueue.size()}"
+  def bufferQueueInfo(): String = s"$this used buffers: ${bufferQueues.map(_.size()).toList}"
 
   override def hashCode(): Int = {
     workingDir.hashCode()

--- a/server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala
+++ b/server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala
@@ -55,14 +55,6 @@ private[worker] final class DiskFlusher(
   private lazy val diskFlusherId = System.identityHashCode(this)
   private val workingQueues = new Array[LinkedBlockingQueue[FlushTask]](threadCount)
   private val bufferQueues = new Array[LinkedBlockingQueue[CompositeByteBuf]](threadCount)
-  for (index <- 0 until threadCount) {
-    val actualQueueSize = queueCapacity / threadCount + 1
-    workingQueues(index) = new LinkedBlockingQueue[FlushTask](actualQueueSize)
-    bufferQueues(index) = new LinkedBlockingQueue[CompositeByteBuf](actualQueueSize)
-    for (_ <- 0 until actualQueueSize) {
-      bufferQueues(index).put(Unpooled.compositeBuffer(256))
-    }
-  }
   private val workers = new Array[Thread](threadCount)
   private val nextWorkerIndex = new AtomicInteger()
 
@@ -76,7 +68,14 @@ private[worker] final class DiskFlusher(
   init()
 
   private def init(): Unit = {
+    val actualQueueSize = queueCapacity / threadCount + 1
     for (index <- 0 until (threadCount)) {
+      workingQueues(index) = new LinkedBlockingQueue[FlushTask](actualQueueSize)
+      bufferQueues(index) = new LinkedBlockingQueue[CompositeByteBuf](actualQueueSize)
+      for (_ <- 0 until actualQueueSize) {
+        bufferQueues(index).put(Unpooled.compositeBuffer(256))
+      }
+
       workers(index) = new Thread(s"$this-$index") {
         override def run(): Unit = {
           while (!stopFlag.get()) {
@@ -117,7 +116,7 @@ private[worker] final class DiskFlusher(
 
   def getWorkerIndex: Int = {
     val nextIndex = nextWorkerIndex.getAndIncrement()
-    if (nextIndex > 1000000) {
+    if (nextIndex > threadCount) {
       nextWorkerIndex.set(0)
     }
     nextIndex % threadCount

